### PR TITLE
Added executing reconcile using the inflight queue

### PR DESCRIPTION
**Pull request by ppatierno**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

This PR adds the periodic reconciliation executed through the inflight queue. Other than being something missing before that, the PR should fix the #258 

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging

### CODE REVIEW

File: topic-controller/src/main/java/io/strimzi/controller/topic/Session.java [hunk 1]
```diff
@@ -166,9 +166,14 @@ private void reconcileTopics(String reconciliationType) {
                     TopicName topicName = new TopicName(name);
                     k8s.getFromName(topicName.asMapName(), ar -> {
                         ConfigMap cm = ar.result();
-                        // TODO need to check inflight
-                        // TODO And need to prevent pileup of inflight periodic reconciliations
-                        controller.reconcile(cm, topicName);
+
+                        controller.reconcile(cm, topicName, reconcileResult -> {
+                            if (reconcileResult.succeeded()) {
+                                LOGGER.info("Success reconciling topic {}", topicName);
```
Comment (tombentley) [inline topic-controller/src/main/java/io/strimzi/controller/topic/Session.java]: You could also add the `reconciliationType` to these log messages, for extra context. Also, is there a reason you're not logging the `cm` name here?

File: topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java [hunk 2]
```diff
@@ -939,5 +957,9 @@ public boolean isWorkInflight() {
     private BackOff topicMetadataBackOff() {
         return new BackOff(config.get(Config.TOPIC_METADATA_MAX_ATTEMPTS));
     }
+
+    static String logConfigMap(ConfigMap cm) {
```
Comment (tombentley) [inline topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java]: Javadoc?

File: topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java [hunk 3]
```diff
@@ -334,30 +334,48 @@ public Controller(Vertx vertx, Kafka kafka,
...
                 }
-            });
-        } catch (InvalidConfigMapException e) {
-            LOGGER.error("Error reconciling ConfigMap {}: Invalid 'data' section: ", cm.getMetadata().getName(), e.getMessage());
-        } catch (ControllerException e) {
-            LOGGER.error("Error reconciling ConfigMap {}: ", cm.getMetadata().getName(), e);
-        }
+            }
+        };
+        inFlight.enqueue(topicName, resultHandler, futureHandler);
```
Comment (tombentley) [inline topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java]: I think we should rename `futureHandler` → `action` or something to make clear that it's the piece of work to be run. I think we should swap the order of the parameters in `enqueue()`, so the action goes before the result handler. That would make it more consistent with the Vertx APIs, like `executeBlocking()`.

File: topic-controller/src/main/java/io/strimzi/controller/topic/Session.java (hunk 1 above)
Comment (ppatierno) [reply topic-controller/src/main/java/io/strimzi/controller/topic/Session.java]: No reasons I'm going to add both.

File: topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java (hunk 2 above)
Comment (ppatierno) [reply topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java]: Done

File: topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java (hunk 3 above)
Comment (ppatierno) [reply topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java]: Agree, do we want do that in this PR ? Maybe a different one involving the other methods could be better, wdyt ?

**Comment by ppatierno** [comment]
@tombentley you are right, waiting for @serrss to test it. You mentioned the possibility to test it but taking a look at the log, I see that it appens when a periodic reconciliation starts just after a ConfigMap is added (in the case of Kafka Connect, more topics are created). It should be something that you already mentioned (talking about the cluster controller ?) but right now we don't have a way for triggering a reconcile on a topic controller instance from unit tests (at least). Or I'm missing something.

File: topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java (hunk 3 above)
Comment (tombentley) [reply topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java]: I think it's a trivial refactoring in the IDE, so we can just do it here.

**Comment by tombentley** [comment]
>  Or I'm missing something.

I think you're missing something :-) In `ControllerIT.setup()` we create a `Session` instance and deploy it through vertx. Session has a `private void reconcileTopics(String)` method which you can make package access. Then in the `ControllerIT` make that session into a field, rather than a local var in `setup()`. Then you can write a test where you can call `reconcileTopics()` directly. I think that should be enough, no?

**Comment by ppatierno** [comment]
I was right ! I missed something ;)

File: topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java (hunk 3 above)
Comment (ppatierno) [reply topic-controller/src/main/java/io/strimzi/controller/topic/Controller.java]: Yes that's true. Just done.

**Comment by ppatierno** [comment]
@tombentley I added the test, when you have time for another review pass ;)

**Pull request merged**